### PR TITLE
Add mode to truncate

### PR DESCRIPTION
**Pull request by dubiousjim**
This is part of a series of PRs towards #4017.

This PR adds a `mode` option for truncate. As with mkdir and the makeTemp*
functions, this argument is only used when a new file is created.
Defaults to 0o666.

We make the tests verify that the `mode` used is subject to the
process's umask (here just assumed to be the default 0o022 on Unix, or 0
on Windows).

The comments say "error to specify mode when create is set to `false`";
but the create option is only introduced in a later commit.

### CODE REVIEW

**Comment by dubiousjim** [comment]
I need to (and will) add a unit test for this new functionality.

**Comment by dubiousjim** [comment]
Adding the tests, I realized that the previous implementation of truncate (both before and after this PR) doesn't let one create new files, ... so this PR is idle. Another PR I have queued (haven't pushed yet) does enable truncate to create new files, then the `mode` option will be effective, and the test will work. Can either close this PR, or merge the `create` functionality with it... but there are some things the latter introduces (namely, the `clobber` option) that @ry wanted to consider/discuss first.

File: cli/js/lib.deno.ns.d.ts [hunk 1]
```diff
@@ -1420,20 +1420,36 @@ declare namespace Deno {
...
   /** Synchronously truncates or extends the specified file, to reach the
    * specified `len`.
    *
    *       Deno.truncateSync("hello.txt", 10);
    *
    * Requires `allow-write` permission. */
-  export function truncateSync(name: string, len?: number): void;
+  export function truncateSync(
+    path: string,
+    len?: number,
```
Comment (ry) [inline cli/js/lib.deno.ns.d.ts]: Should len move to the options? cc @piscisaureus

File: cli/ops/fs.rs [hunk 2]
```diff
@@ -694,7 +695,15 @@ fn op_truncate(
   let is_sync = args.promise_id.is_none();
   blocking_json(is_sync, move || {
     debug!("op_truncate {} {}", path.display(), len);
-    let f = fs::OpenOptions::new().write(true).open(&path)?;
+    let mut open_options = fs::OpenOptions::new();
+    open_options.write(true);
+    if let Some(_mode) = args.mode {
```
Comment (ry) [inline cli/ops/fs.rs]: Don't prefix with underscore, just use `mode`.

File: cli/js/truncate.ts [hunk 3]
```diff
@@ -1,6 +1,14 @@
 // Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
 import { sendSync, sendAsync } from "./ops/dispatch_json.ts";
 
+export interface TruncateOptions {
+  /** Permissions to use if creating the file (defaults to `0o666`, before
+   * the process's umask).
+   * It's an error to specify mode when create is set to `false`.
+   * Does nothing/raises on Windows. */
+  mode?: number;
+}
```
Comment (ry) [inline cli/js/truncate.ts]: Ideally we'd add a unit test of this too...

Comment (dubiousjim) [reply cli/js/truncate.ts]: Right, but see my comment above... I realized after pushing this it couldn't be tested until the `create` option is added to truncate too (I was going to do that after you all had a chance to review this permission-related batch of PRs).

File: cli/ops/fs.rs (hunk 2 above)
Comment (dubiousjim) [reply cli/ops/fs.rs]: That's tricky. As implemented, that variable is used only on unix. I had to add the underscore to make the code pass the lint check on windows. You'd think I could just move the `#[cfg(unix)]` to before the `if let...` block, but no, Rust currently doesn't let you put attributes on `if` blocks. What I could do is surround the whole if block in `{ ... }` and then move the `#[cfg(unix)]` to apply to that larger block. Then Windows would never see the `if let(_mode)...` and we could drop the underscore. Do you prefer that? (It seemed 6 of one, half-dozen of the other to me.)

**Comment by dubiousjim** [comment]
Am closing for now because this shouldn't be merged until the `create` option is added. The implementation for that is ready, but I'm waiting for other PRs to be merged, and some other questions to be discussed, before pushing it. When I do push it, I'll re-open this PR and add it here.

**Pull request closed without merging**